### PR TITLE
Transmux overlapping

### PR DESCRIPTION
**Pull request by wlozano0collabora**
transmuxer: Avoid muxing overlapping files

### CODE REVIEW

File: hwangsae/transmuxer.h [hunk 1]
```diff
@@ -25,6 +25,8 @@
 
 #include <glib-object.h>
 
+#include "types.h"
+
```
Comment (justinjoy) [inline hwangsae/transmuxer.h]: This change seems not to be necessary for header. Shouldn't we move `types.h` to `.c` file?

**Review by justinjoy — CHANGES_REQUESTED** [review]

File: hwangsae/transmuxer.c [hunk 2]
```diff
@@ -257,6 +296,13 @@ hwangsae_transmuxer_merge (HwangsaeTransmuxer * self, GSList * input_files,
   g_return_if_fail (input_files != NULL);
   g_return_if_fail (output != NULL);
 
+  if (!_check_time_overlap (input_files)) {
+    g_warning ("There are overlaping segments");
```
Comment (xhaakon) [inline hwangsae/transmuxer.c]: Typo in overlapping

File: hwangsae/transmuxer.c [hunk 3]
```diff
@@ -206,6 +208,43 @@ hwangsae_transmuxer_clear (HwangsaeTransmuxer * self)
   gst_clear_object (&priv->pipeline);
 }
 
+static gboolean
+_check_time_overlap (GSList * input_files)
```
Comment (xhaakon) [inline hwangsae/transmuxer.c]: I guess this check could get simplified:
* add `guint64 duration` to `Segment` structure
* calculate `duration` from start and end times in `hwangsae_transmuxer_parse_segments()`
* then `segments` in `HwangsaeTransmuxerPrivate` will hold the list of input files ordered by start time
* now detect overlaps by iterating `segments` and checking that `base_time + duration < next_segment_base_time`

File: hwangsae/transmuxer.h (hunk 1 above)
Comment (wlozano0collabora) [reply hwangsae/transmuxer.h]: I added it to allow to reference HWANGSAE_TRANSMUXER_ERROR_OVERLAP from application without including, similar approach to the one used in `recorder.h`.

Do you prefer `types.h` be included in application?

File: hwangsae/transmuxer.c (hunk 2 above)
Comment (wlozano0collabora) [reply hwangsae/transmuxer.c]: Thanks.

File: hwangsae/transmuxer.c (hunk 3 above)
Comment (wlozano0collabora) [reply hwangsae/transmuxer.c]: Thanks for the suggestion. I can move the check to `hwangsae_transmuxer_parse_segments`, but  also 
- move `hwangsae_transmuxer_parse_segments` to the beginning of perform this check earlier.
- I don't see the reason to calculate `duration` as you need segment_end_time which is already and output of the `hwangsae_common_parse_times_from_filename`.

Do you agree?

Comment (xhaakon) [reply hwangsae/transmuxer.c]: > move `hwangsae_transmuxer_parse_segments` to the beginning of perform this check earlier.

Yes, nothing against that.

> you need segment_end_time which is already and output of the `hwangsae_common_parse_times_from_filename`.

Yeah that should work too.

File: hwangsae/transmuxer.c [hunk 4]
```diff
@@ -215,31 +218,42 @@ hwangsae_transmuxer_parse_segments (HwangsaeTransmuxer * self,
...
+    segment->end_time = segment_end;
 
     segments = g_list_insert_sorted (segments, segment, _compare_segments);
   }
 
+
+  base_time_prev = 0;
   /* Make base times start from zero. */
   for (it = g_list_last (segments); it; it = it->prev) {
+    if (base_time_prev != 0
```
Comment (xhaakon) [inline hwangsae/transmuxer.c]: Hmm, for the sake of clarity I'd rather move the overlap check to `hwangsae_transmuxer_merge()`.

File: hwangsae/transmuxer.c [hunk 5]
```diff
@@ -219,16 +222,18 @@ hwangsae_transmuxer_parse_segments (HwangsaeTransmuxer * self,
...
+    if (!hwangsae_common_parse_times_from_filename (file, &segment_start,
+            &segment_end)) {
       g_warning ("Invalid filename %s", file);
       continue;
     }
 
     segment = g_new0 (Segment, 1);
     segment->filename = g_strdup (file);
     segment->base_time = segment_start;
+    segment->end_time = segment_end;
```
Comment (xhaakon) [inline hwangsae/transmuxer.c:327]: You should subtract first segment's `base_time` from end times as well in the for loop below.

File: hwangsae/transmuxer.c [hunk 6]
```diff
@@ -242,6 +247,22 @@ hwangsae_transmuxer_parse_segments (HwangsaeTransmuxer * self,
...
+      hwangsae_transmuxer_get_instance_private (self);
+  GList *it;
+  gint64 end_time_prev = 0;
+
+  for (it = g_list_first (priv->segments); it; it = it->next) {
+    if (end_time_prev != 0 && ((Segment *) it->data)->base_time < end_time_prev)
+      return FALSE;
+    end_time_prev = ((Segment *) it->data)->end_time;
+  }
+  return TRUE;
```
Comment (xhaakon) [inline hwangsae/transmuxer.c]: This should work, but maybe we could take advantage of `segments` being a doubly-linked list and write the check more elegantly, e.g.

```c
HwangsaeTransmuxerPrivate *priv =
    hwangsae_transmuxer_get_instance_private (self);
GList *it;

if (!priv->segments) {
  /* Empty segment list. */
  return TRUE;
}

for (it = priv->segments->next; it; it = it->next) {
  Segment *segment = it->data;
  Segment *previous_segment = it->prev->data; 

  if (previous_segment->end_time > segment->base_time)
    return FALSE;
  }
}

return TRUE;
```

File: hwangsae/transmuxer.h (hunk 1 above)
Comment (justinjoy) [reply hwangsae/transmuxer.h]: I think it's better to keep `.h` files simple. When I moved this statement to `.c`, it was built successfully. 

And if you want to keep `types.h` for public header, it should be `<types.h>` for consistency.

Comment (wlozano0collabora) [reply hwangsae/transmuxer.h]: Thanks for your time and review. You are right, as `types.h` is in `hwangsae.h` it is not necessary here as I initially supposed.

File: hwangsae/transmuxer.c (hunk 5 above)
Comment (wlozano0collabora) [reply hwangsae/transmuxer.c:327]: Yes, you are right.

File: hwangsae/transmuxer.c (hunk 6 above)
Comment (wlozano0collabora) [reply hwangsae/transmuxer.c]: Yes, it looks much better, thanks for your suggestion.

**Review by xhaakon — COMMENTED** [review]
LGTM, but needs a rebase.

**Comment by wlozano0collabora** [comment]
Sure @xhaakon, that is not a problem. Thanks for your time.

**Comment by wlozano0collabora** [comment]
This rebase will be applied after merging #70

**Review by xhaakon — APPROVED** [review]

**Review by justinjoy — APPROVED** [review]
thank you!

**Pull request merged**